src/main/java/org/broadinstitute/hellbender/tools/longreads/graph/AlignedBaseGraphCollection.java

```diff
@@ -0,0 +1,461 @@
+        contigPositionVertexMap.put(contig, nodePositionMap);
+
+        AlignedBaseVertex lastVertex = null;
+        AlignedBaseVertex vertex;
+
+        while (nodes.peek() != null) {
+
+            // Get our vertex:
+            vertex = nodes.remove();
+
+            // Add the vertex to our graph
+            graph.addVertex(vertex);
+
+            // Add the link in the graph from the last vertex to this one:
+            if ( lastVertex != null ) {
+                graph.addEdge(lastVertex, vertex);
+            }
+
+            // Add the nodes to our node store:
+            if (contigPositionVertexMap.get(contig).containsKey(vertex.getPos())) {
+                contigPositionVertexMap.get(contig).get(vertex.getPos()).add(vertex);
+            }
+            else {
+                contigPositionVertexMap.get(contig).put(
+                        vertex.getPos(),
+                        new HashSet<>(Collections.singletonList(vertex))
+                );
+            }
+
+            // Store the last vertex:
+            lastVertex = vertex;
+        }
+
+        // Add the graph to our contig graph map:
+        contigSubGraphMap.put(contig, graph);
+    }
+
+    /**
+     * Adds all given {@code nodes} into {@link #contigSubGraphMap}.
+     * Assumes that {@code contig} exists in {@link #contigSubGraphMap} and {@link #contigPositionVertexMap}.
+     * @param contig Contig from which the given nodes originate.
+     * @param nodes {@link Queue<AlignedBaseVertex>} representing a linear graph of bases in a given read.
+     */
+    private void mergeNodesIntoGraph(final String contig, final Queue<AlignedBaseVertex> nodes) {
+        // Get the graph we'll be working with:
+        final AlignedBaseGraph graph = contigSubGraphMap.get(contig);
+
+        // Get the position map we'll be dealing with:
+        final TreeMap<GenomicAndInsertionPosition, Set<AlignedBaseVertex>> positionVertexMap =
+                contigPositionVertexMap.get(contig);
+
+        AlignedBaseVertex lastVertex = null;
+        AlignedBaseVertex vertex;
+
+        boolean lastNodeWasAddedToGraph = false;
+
+        // Iterate through our nodes and find the best place to insert them:
+        while (nodes.peek() != null) {
+            vertex = nodes.remove();
+
+            // Get the nodes in the graph that aligns to this one:
+            final Set<AlignedBaseVertex> graphVertices = positionVertexMap.get(vertex.getPos());
+
+            // Does any graph node have the same allele info as the current node?
+            // If so, we don't add a new vertex to the graph and we set the vertex to the graph vertex to
+            // preserve link continuity later:
+            boolean vertexIsInGraph = false;
+            if ( graphVertices != null ) {
+                for ( final AlignedBaseVertex graphVertex : graphVertices ) {
+                    if ( vertex.isSequenceEqual(graphVertex.getSequence()) ) {
```

Comment:
I think this can cause problems when you have multiple (different) insertions at a site. Specifically Read A has something like ....AA(CCAACCAA)GGGG... and read B has ...AA(CCCCC)GGGG... (where the stuff in brackets are insertions relative to the reference) since each base is keyed by position -> insertion offset then you might cause problems where the two different insertions at the same site get remerged with eachother and information is lost. This might just need to be a known issue as it is a fairly edge case. It would seem to imply there is a bizzare left preference for insertions, but only overlapping ones. Around STR regions it seems possible to wind up with the same indel being represented at multiple sites in the graph, it seems prudent to perhaps leftalign indels on your reads to collapse that complexity. 
These issues (among other complexities inside of insertions) do seem particularly painful if you are expecting to capture reasonable behavior in SV or CNV events in your long read data though.... Its a complex problem to solve.

Reply:
I don't think this is an issue - each vertex is a single aligned base, so if both the aligned base position and the base itself is not equal, it will create a new path in the graph.  If they are equal then that existing path is used.  

@jamesemery for your case the graph would end up looking like:
```
      |------ insertion ------|

          /->A->A-\
A->A->C->C         ->C->C->A->A->G->G->G->G
          \->C->C-/
```

If the reads are all perfectly mapped, then this is not an issue (I have to make the assumption that the mapping is error-free to make a sensical graph).  If the mapping has no errors, then different indels in an STR would resolve to different positions and would be represented differently.